### ISSUE TITLE
Pillarenv doesn't allow using separate pillar environments

### ISSUE THREAD

**Issue by krcroft** [saltstack/salt#36153]
### Description of Issue/Question

Pillarenv, as documented, describes the ability to select a specific pillar environment at runtime using the pillarenv=<environment> argument, however Salt ignores this and sequentially merges all the environments defined in the salt/master config.  This multi-environment merged pillar is actually what's used.
### Steps to Reproduce Issue

/etc/salt/master:

```
pillar_roots:
  a:
    - /srv/a/pillar
  b:
    - /srv/b/pillar

```

/srv/*/pillar/top.sls:

```
a:
  '*':
    - a
b:
  '*':
    - b
```

/srv/*/pillar/a.sls:
`value: a`

/srv/*/pillar/b.sls:
`value: b`
### Expected behavior

`salt node pillar.item value pillarenv=a`

Or

Salt node state.apply some/state pillarenv=a

The expected pillar value is a, but actual is b.
### Versions Report

> [root@progen /scratch/build] salt --versions-report
> Salt Version:
>            Salt: 2016.3.3
> 
> Dependency Versions:
>            cffi: 1.7.0
>        cherrypy: Not Installed
>        dateutil: 2.5.3
>           gitdb: Not Installed
>       gitpython: Not Installed
>           ioflo: Not Installed
>          Jinja2: 2.8
>         libgit2: Not Installed
>         libnacl: Not Installed
>        M2Crypto: Not Installed
>            Mako: Not Installed
>    msgpack-pure: Not Installed
>  msgpack-python: 0.4.7
>    mysql-python: Not Installed
>       pycparser: 2.14
>        pycrypto: 2.6.1
>          pygit2: Not Installed
>          Python: 2.6.6 (r266:84292, Aug 18 2016, 15:13:37)
>    python-gnupg: Not Installed
>          PyYAML: 3.11
>           PyZMQ: 15.2.0
>            RAET: Not Installed
>           smmap: Not Installed
>         timelib: 0.2.4
>         Tornado: 4.3
>             ZMQ: 4.1.2

**Comment by Ch3LL** [comment]
@krcroft I believe this is a duplicate of #32435 . Are you okay closing this and monitoring this issue there? thanks

**Comment by krcroft** [comment]
Hi Ch3LL,

No.  I looked at the proposed duplicate issue, and it's not the same. This issue here discusses improper operation of the pillarenv feature, where as the other issue discusses how pillarenv is missing from some salt states/functions/modules.

**Comment by Ch3LL** [comment]
@krcroft oh i see it looks like you are also stating that when running state.apply with the pillarenv parameter its still not working. I am not able to reproduce this and will provide a docker container below:

1. `docker run -it -v /home/ch3ll/git/salt/:/testing/ ch3ll/issues:36153 /bin/bash` (where /home/ch3ll/git/salt/ is a local cloned git repo of salt)
2. `salt-master -d; salt-minion -d`
3. run the state file: `salt '*' state.apply test pillarenv=b -lwarning`

As you can see below it seems `pillarenv` is correclty switching the pillarenv and my state produces `echo b` instead of `echo a`:

```
694981d59e09:
----------
          ID: run_cmd
    Function: cmd.run
        Name: echo b
      Result: True
     Comment: Command "echo b" run
     Started: 21:59:24.461436
    Duration: 14.481 ms
     Changes:   
              ----------
              pid:
                  387
              retcode:
                  0
              stderr:
              stdout:
                  b

Summary for 694981d59e09
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  14.481 ms
```

vs `pillarenv` a:

```
[root@c7f60067f9da /]# salt '*' state.apply test pillarenv=a -lwarning
694981d59e09:
----------
          ID: run_cmd
    Function: cmd.run
        Name: echo a
      Result: True
     Comment: Command "echo a" run
     Started: 22:01:49.207151
    Duration: 13.302 ms
     Changes:   
              ----------
              pid:
                  428
              retcode:
                  0
              stderr:
              stdout:
                  a

Summary for 694981d59e09
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  13.302 ms
```

And here is my state:

```
[root@c7f60067f9da /]# cat /srv/salt/test.sls 
run_cmd:
  cmd.run:
    - name: echo {{ salt.pillar.get('value', '') }}
```
 
Is there something i'm missing from my test case to help reproduce the `state.apply` issue?

As far as the `pillar.item` not working when attempting to call a differnet pillarenv I still believe its related to that issue because specifying a different pillarenv has not been added to that module call `pillar.item`.

**Comment by cmclaughlin** [comment]
I've been bothered by this too... given the example above:

if you delete b.sls and use pillarenv=b, won't "echo a" run?  In my opinion when you specify saltenv and pillarenv, no merging should happen.  Perhaps there should be a setting for this.

**Comment by Ch3LL** [comment]
So if I delete b.sls and proof shown below:

```
[root@98d948f4ce76 pillar]# salt '*' pillar.items -lwarning
694981d59e09:
    ----------
    _errors:
        - Specified SLS 'b' in environment 'b' is not available on the salt master
    value:
        a
```

it returns nothing which I would expect instead of a when specifying pillarenv:

```
[root@98d948f4ce76 pillar]# salt '*' state.apply test pillarenv=b -lwarning
694981d59e09:
----------
          ID: run_cmd
    Function: cmd.run
        Name: echo
      Result: True
     Comment: Command "echo" run
     Started: 22:56:58.765533
    Duration: 13.908 ms
     Changes:   
              ----------
              pid:
                  853
              retcode:
                  0
              stderr:
              stdout:

Summary for 694981d59e09
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  13.908 ms
```

@cmclaughlin are you seeing a case where this does not occur

i'm still not seeing where the error persists in this issue. I can't seem to replicate it. If anyone can see what i'm doing wrong?

**Comment by cmclaughlin** [comment]
@Ch3LL yeah I can't reproduce this either as described in this ticket.  I've definitely had some frustrations with environments, but I might not be able to remember the exact problems.  I guess this wasn't it. Thanks!

**Comment by Ch3LL** [comment]
@cmclaughlin if you ever remember feel free to open an issue and we can track it there thanks

**Comment by rallytime** [comment]
Closed via #46543

**Comment by krcroft** [comment]
@Ch3LL, @rallytime  -- thanks!
